addons/binding/org.openhab.binding.dlinksmarthome/src/main/java/org/openhab/binding/dlinksmarthome/internal/DLinkHNAPCommunication.java

```diff
@@ -0,0 +1,437 @@
+
+                action.saveChanges();
+            } catch (final InvalidKeyException e) {
+                logger.debug("setAuthenticationHeaders - Internal error", e);
+                status = HNAPStatus.INTERNAL_ERROR;
+            } catch (final NoSuchAlgorithmException e) {
+                logger.debug("setAuthenticationHeaders - Internal error", e);
+                status = HNAPStatus.INTERNAL_ERROR;
+            } catch (final SOAPException e) {
+                // No communication happening so assume system error
+                logger.debug("setAuthenticationHeaders - Internal error", e);
+                status = HNAPStatus.INTERNAL_ERROR;
+            }
+        }
+    }
+
+    /**
+     * Send the SOAP message using a HttpURLConnection and parse the result using DocumentBuilder.
+     * This is used in preference to SOAPConnection and SOAPMessage as these tend to fill
+     * the log file with messages when things go wrong.
+     *
+     * @param action - SOAP Action to send
+     * @param timeout - Connection timeout in milliseconds
+     * @return The result which may be null
+     * @throws IOException
+     * @throws SOAPException
+     * @throws SAXException
+     */
+    protected Document sendReceive(final SOAPMessage action, final int timeout)
+            throws IOException, SOAPException, SAXException {
+
+        Document result = null;
+
+        final HttpURLConnection connection = (HttpURLConnection) url.openConnection();
+
+        connection.setConnectTimeout(timeout);
+        connection.setReadTimeout(timeout);
+
+        connection.setDoOutput(true);
+        final Iterator<?> it = action.getMimeHeaders().getAllHeaders();
+        while (it.hasNext()) {
+            final MimeHeader header = (MimeHeader) it.next();
+            connection.setRequestProperty(header.getName(), header.getValue());
+        }
+
+        try (final OutputStream os = connection.getOutputStream()) {
+            action.writeTo(os);
+        }
+
+        try (final InputStream is = connection.getInputStream()) {
+            result = parser.parse(is);
+        }
+
+        return result;
+    }
+
+    /**
+     * Output unexpected responses to the debug log.
+     *
+     * @param message
+     * @param soapResponse
+     */
+    protected void logUnexpectedResult(final String message, final Document soapResponse) {
+
+        try {
+            final TransformerFactory transFactory = TransformerFactory.newInstance();
+            final Transformer transformer = transFactory.newTransformer();
+            final StringWriter buffer = new StringWriter();
+            transformer.setOutputProperty(OutputKeys.OMIT_XML_DECLARATION, "yes");
+            transformer.transform(new DOMSource(soapResponse), new StreamResult(buffer));
+            logger.debug("{} : {}", message, buffer.toString());
+        } catch (final TransformerException e) {
+            logger.debug("{}", message);
```

Comment:
By this stage I would have completely lost the response message I was trying to log so I don't think this exception is of any use.